### PR TITLE
Redshift data sharing - Cluster encryption guardrails and information

### PR DESCRIPTION
**Pull request by dlpzx**
### Feature or Bugfix
- Feature
 TO BE MERGED AFTER https://github.com/data-dot-all/dataall/pull/1424 and after #1449 

### Detail

This PR is part of the design explained in detail in #955 
- Add check that verifies the Redshift cluster imported is encrypted before creating a redshift connection
- Store the Redshift encryption type in the redshift-connection table. This is needed because Redshift datashares only work between namespaces that use the same encryption type
- In the migration script and in the db model of the redshift-connection table we add the columns encryptionType and connectionType. ConnectionType is not needed in this particular PR, but it will be needed in the next PR. It is added to simplify the migration scripts

### Relates
- #955 - Sharing part

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)?
  - Is the input sanitized?
  - What precautions are you taking before deserializing the data you consume?
  - Is injection prevented by parametrizing queries?
  - Have you ensured no `eval` or similar functions are used?
- Does this PR introduce any functionality or component that requires authorization?
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms?
  - Are you logging failed auth attempts?
- Are you using or adding any cryptographic features?
  - Do you use a standard proven implementations?
  - Are the used keys controlled by the customer? Where are they stored?
- Are you introducing any new policies/roles/users?
  - Have you used the least-privilege principle? How?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by dlpzx** [comment]
Testing:
- (no testing needed for this scenario) - Serverless clusters are encrypted by default, so it is not possible for them to not be encrypted. At a minimum they are encrypted with AWS-Managed keys
- [ ]  Provisioned clusters not encrypted cannot be onboarded with data.all Redshift connections - creation of a connection throws an error
- [X] Serverless clusters encrypted with customer managed keys are correctly classified as `RedshiftEncryptionType.CUSTOMER_MANAGED_KMS_KEY`
- [ ] Serverless clusters encrypted with AWS managed keys are correctly classified as `RedshiftEncryptionType.AWS_OWNED_KMS_KEY`
- [ ] Provisioned clusters encrypted with customer managed keys are correctly classified as `RedshiftEncryptionType.CUSTOMER_MANAGED_KMS_KEY`
- [ ] Provisioned clusters encrypted with AWS managed keys are correctly classified as `RedshiftEncryptionType.AWS_OWNED_KMS_KEY`
- [ ] Provisioned clusters encrypted with HSM are correctly classified as `RedshiftEncryptionType.HSM`

File: backend/dataall/modules/redshift_datasets/services/redshift_connection_service.py [hunk 1]
```diff
@@ -46,10 +52,14 @@ def create_redshift_connection(uri, admin_group, data: dict) -> RedshiftConnecti
                 database=data.get('database'),
                 redshiftUser=data.get('redshiftUser', ''),
                 secretArn=data.get('secretArn', ''),
+                connectionType=data.get('connectionType', RedshiftConnectionTypes.DATA_USER.value),
```
Comment (noah-paige) [inline backend/dataall/modules/redshift_datasets/services/redshift_connection_service.py:55]: does not look like we pass a parameter for `connectionType` - so will every connectionType be defaulted to `RedshiftConnectionTypes.DATA_USER.value`?

File: backend/dataall/modules/redshift_datasets/services/redshift_connection_service.py [hunk 2]
```diff
@@ -161,3 +174,30 @@ def _check_redshift_connection(account_id: str, region: str, connection: Redshif
...
+    def _get_redshift_encryption(
+        account_id: str, region: str, connection: RedshiftConnection
+    ) -> RedshiftEncryptionType:
+        if connection.redshiftType == RedshiftType.Serverless.value:
+            namespace = redshift_serverless_client(account_id=account_id, region=region).get_namespace_by_id(
+                connection.nameSpaceId
+            )
+            return (
+                RedshiftEncryptionType.AWS_OWNED_KMS_KEY
+                if namespace.get('kmsKeyId', None) == RedshiftEncryptionType.AWS_OWNED_KMS_KEY.value
```
Comment (noah-paige) [inline backend/dataall/modules/redshift_datasets/services/redshift_connection_service.py:188]: do we not need to `describe_kms_key` similar to how we do with Provisioned Clusters L194 below?

File: backend/dataall/modules/redshift_datasets/services/redshift_connection_service.py [hunk 3]
```diff
@@ -161,3 +174,30 @@ def _check_redshift_connection(account_id: str, region: str, connection: Redshif
...
+                key = kms_redshift_client(account_id=account_id, region=region).describe_kms_key(key_id=key_id)
+                if key.get('KeyManager', None) == 'AWS':
+                    return RedshiftEncryptionType.AWS_OWNED_KMS_KEY
+                elif key.get('KeyManager', None) == 'CUSTOMER':
+                    return RedshiftEncryptionType.CUSTOMER_MANAGED_KMS_KEY
+                else:
+                    raise Exception
+            if cluster.get('HsmStatus', None):
+                return RedshiftEncryptionType.HSM
+        raise Exception
```
Comment (noah-paige) [inline backend/dataall/modules/redshift_datasets/services/redshift_connection_service.py:203]: nit: raise generic Exceptions with no error message

**Review by noah-paige — APPROVED** [review]
lgtm - approving

left one minor comment consider it optional and can pick up in future PR part

**Pull request merged**